File: explainability_app/wire/responses.py
"""Pydantic response shapes for the explainability API.

These are the wire-level contracts consumed by the frontend Zod mirrors.
The `__python_schema_baseline__.json` file in the frontend mirrors a
JSON-Schema export of every shape defined here; baseline_drift tests fail
on any unsynchronised change.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class WorkflowSummaryResponse(BaseModel):
    workflow_id: str
    started_at: datetime | None = None
    event_count: int = 0
    status: str = "unknown"
    primary_agent_id: str | None = None


class BlackBoxEventResponse(BaseModel):
    event_id: str
    workflow_id: str
    event_type: str
    timestamp: datetime | None = None
    step: int | None = None
    details: dict[str, Any] = {}
    integrity_hash: str = ""


class WorkflowEventsResponse(BaseModel):
    workflow_id: str
    event_count: int = 0
    hash_chain_valid: bool = False
    events: list[BlackBoxEventResponse] = []


class DecisionRecordResponse(BaseModel):
    workflow_id: str
    phase: str
    description: str
    alternatives: list[str] = []
    rationale: str
    confidence: float
    timestamp: datetime | None = None


class TimeSeriesPointResponse(BaseModel):
    bucket: datetime
    value: float


class DashboardMetricsResponse(BaseModel):
    total_runs: int = 0
    p50_latency_ms: float = 0.0
    p95_latency_ms: float = 0.0
    total_cost_usd: float = 0.0
    guardrail_pass_rate: float = 0.0
    hash_chain_valid_count: int = 0
    hash_chain_invalid_count: int = 0
    time_series_cost: list[TimeSeriesPointResponse] = []
    time_series_latency: list[TimeSeriesPointResponse] = []
    time_series_tokens: list[TimeSeriesPointResponse] = []
    model_distribution: dict[str, int] = {}


class ValidatorStatResponse(BaseModel):
    name: str
    total_checks: int = 0
    pass_count: int = 0
    fail_count: int = 0
    pass_rate: float = 0.0


class GuardrailFailureResponse(BaseModel):
    workflow_id: str
    validator: str
    fail_action: str | None = None
    timestamp: datetime | None = None


class GuardrailSummaryResponse(BaseModel):
    total_checks: int = 0
    pass_count: int = 0
    fail_count: int = 0
    pass_rate: float = 0.0
    fail_action_distribution: dict[str, int] = {}
    per_validator: list[ValidatorStatResponse] = []
    recent_failures: list[GuardrailFailureResponse] = []
    trend_pass_rate_delta: float = 0.0


class CapabilityResponse(BaseModel):
    name: str
    description: str = ""
    parameters: dict[str, Any] = {}


class PolicyResponse(BaseModel):
    name: str
    description: str = ""
    rules: dict[str, Any] = {}


class AgentCardResponse(BaseModel):
    """Wire shape for the read-only Agent Registry card.

    ``signature_verification_status`` distinguishes the three trust signals
    (``verified``, ``failed``, ``unavailable``); ``signature_verified`` is
    retained as a convenience boolean (``True`` only for ``verified``) so
    older clients keep working while the UI migrates to the richer status
    string.
    """

    agent_id: str
    agent_name: str
    owner: str
    version: str
    description: str = ""
    capabilities: list[CapabilityResponse] = []
    policies: list[PolicyResponse] = []
    status: str
    valid_until: datetime | None = None
    parent_agent_id: str | None = None
    signature_truncated: str
    signature_verified: bool
    signature_verification_status: str = "unavailable"
    created_at: datetime
    updated_at: datetime


class AgentAuditEntryResponse(BaseModel):
    agent_id: str
    action: str
    performed_by: str
    timestamp: datetime
    details: dict[str, Any] = {}


class IntegrityReportResponse(BaseModel):
    """S3.1.1 wire shape: chain-integrity result for a single workflow."""

    workflow_id: str
    chain_valid: bool
    broken_at_event_id: str | None = None
    expected_hash: str | None = None
    actual_hash: str | None = None


class CorrelationHealthResponse(BaseModel):
    """S3.1.2 wire shape: explicit per-key correlation status.

    `missing_keys` is the convenience list consumed by the Compliance UI's
    "missing keys named explicitly" card.
    """

    has_trace_id: bool
    has_user_id: bool
    has_task_id: bool
    has_agent_id: bool
    missing_keys: list[str] = []


class ComplianceBundleResponse(BaseModel):
    """S3.1.2 wire shape: four-pillar compliance bundle for a workflow.

    The ``integrity`` field embeds the full `IntegrityReportResponse` so the
    Workflow Deep Dive's Recording quadrant can name the broken event id
    plus the expected/actual hashes (Sprint 3 review F4).
    """

    workflow_id: str
    event_count: int = 0
    hash_chain_valid: bool = False
    bundle_type: str = "compliance_audit"
    exported_at: datetime | None = None
    events: list[BlackBoxEventResponse] = []
    identity_cards: dict[str, AgentCardResponse | None] = {}
    audit_trails: dict[str, list[AgentAuditEntryResponse]] = {}
    phase_decisions: list[DecisionRecordResponse] = []
    correlation_health: CorrelationHealthResponse
    integrity: IntegrityReportResponse


class WorkflowIntegritySummaryResponse(BaseModel):
    """One row in the batched compliance summary.

    Pairs a workflow summary with its integrity report (or ``None`` when
    the workflow could not be read).  See `ComplianceSummaryResponse`.
    """

    workflow: WorkflowSummaryResponse
    integrity: IntegrityReportResponse | None = None


class ComplianceSummaryResponse(BaseModel):
    """Batched compliance home payload (Sprint 3 review F3 fix).

    Replaces the previous per-workflow integrity fan-out: one call returns
    every workflow + integrity row, plus the bounds the server actually
    applied (so the UI can render the audit window verbatim).
    """

    rows: list[WorkflowIntegritySummaryResponse] = []
    generated_at: datetime
    since: datetime | None = None
    until: datetime | None = None


class LogRowResponse(BaseModel):
    """S4.3.1 wire shape: a single parsed log line.

    `level="UNKNOWN"` when the formatter pattern did not match -- the raw line
    is preserved in `raw` and `message` so the operator can still see the
    bytes that hit disk.
    """

    concern: str
    timestamp: datetime | None = None
    logger: str = ""
    level: str = "UNKNOWN"
    message: str = ""
    raw: str = ""


class HealthResponse(BaseModel):
    status: str = "ok"


class ErrorResponse(BaseModel):
    detail: str

File: frontend-explainability/lib/wire/responses.ts
/**
 * Wire-layer Zod schemas — mirrors of explainability_app/wire/responses.py.
 *
 * Rule W7: each schema is co-exported alongside its inferred TS type.
 * Rule W2: __python_schema_baseline__.json tracks the Python-side JSON Schema
 *          for every shape in this file; baseline_drift.test.ts asserts they
 *          stay in sync (per-shape comparison, fails on any drift).
 *
 * @sdk zod ^3.0.0
 */
import { z } from "zod";

export const WorkflowSummarySchema = z.object({
  workflow_id: z.string(),
  started_at: z.string().datetime({ offset: true }).nullable(),
  event_count: z.number().int().nonnegative(),
  status: z.string(),
  primary_agent_id: z.string().nullable(),
});
export type WorkflowSummary = z.infer<typeof WorkflowSummarySchema>;

export const WorkflowSummaryListSchema = z.array(WorkflowSummarySchema);
export type WorkflowSummaryList = z.infer<typeof WorkflowSummaryListSchema>;

export const BlackBoxEventSchema = z.object({
  event_id: z.string(),
  workflow_id: z.string(),
  event_type: z.string(),
  timestamp: z.string().datetime({ offset: true }).nullable(),
  step: z.number().int().nullable(),
  details: z.record(z.unknown()),
  integrity_hash: z.string(),
});
export type BlackBoxEvent = z.infer<typeof BlackBoxEventSchema>;

export const WorkflowEventsSchema = z.object({
  workflow_id: z.string(),
  event_count: z.number().int().nonnegative(),
  hash_chain_valid: z.boolean(),
  events: z.array(BlackBoxEventSchema),
});
export type WorkflowEvents = z.infer<typeof WorkflowEventsSchema>;

export const DecisionRecordSchema = z.object({
  workflow_id: z.string(),
  phase: z.string(),
  description: z.string(),
  alternatives: z.array(z.string()),
  rationale: z.string(),
  confidence: z.number(),
  timestamp: z.string().datetime({ offset: true }).nullable(),
});
export type DecisionRecord = z.infer<typeof DecisionRecordSchema>;

export const DecisionRecordListSchema = z.array(DecisionRecordSchema);
export type DecisionRecordList = z.infer<typeof DecisionRecordListSchema>;

export const TimeSeriesPointSchema = z.object({
  bucket: z.string().datetime({ offset: true }),
  value: z.number(),
});
export type TimeSeriesPoint = z.infer<typeof TimeSeriesPointSchema>;

export const DashboardMetricsSchema = z.object({
  total_runs: z.number().int().nonnegative(),
  p50_latency_ms: z.number(),
  p95_latency_ms: z.number(),
  total_cost_usd: z.number(),
  guardrail_pass_rate: z.number(),
  hash_chain_valid_count: z.number().int().nonnegative(),
  hash_chain_invalid_count: z.number().int().nonnegative(),
  time_series_cost: z.array(TimeSeriesPointSchema),
  time_series_latency: z.array(TimeSeriesPointSchema),
  time_series_tokens: z.array(TimeSeriesPointSchema),
  model_distribution: z.record(z.number().int().nonnegative()),
});
export type DashboardMetrics = z.infer<typeof DashboardMetricsSchema>;

export const ValidatorStatSchema = z.object({
  name: z.string(),
  total_checks: z.number().int().nonnegative(),
  pass_count: z.number().int().nonnegative(),
  fail_count: z.number().int().nonnegative(),
  pass_rate: z.number(),
});
export type ValidatorStat = z.infer<typeof ValidatorStatSchema>;

export const GuardrailFailureSchema = z.object({
  workflow_id: z.string(),
  validator: z.string(),
  fail_action: z.string().nullable(),
  timestamp: z.string().datetime({ offset: true }).nullable(),
});
export type GuardrailFailure = z.infer<typeof GuardrailFailureSchema>;

export const GuardrailSummarySchema = z.object({
  total_checks: z.number().int().nonnegative(),
  pass_count: z.number().int().nonnegative(),
  fail_count: z.number().int().nonnegative(),
  pass_rate: z.number(),
  fail_action_distribution: z.record(z.number().int().nonnegative()),
  per_validator: z.array(ValidatorStatSchema),
  recent_failures: z.array(GuardrailFailureSchema),
  trend_pass_rate_delta: z.number(),
});
export type GuardrailSummary = z.infer<typeof GuardrailSummarySchema>;

export const CapabilitySchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.record(z.unknown()),
});
export type Capability = z.infer<typeof CapabilitySchema>;

export const PolicySchema = z.object({
  name: z.string(),
  description: z.string(),
  rules: z.record(z.unknown()),
});
export type Policy = z.infer<typeof PolicySchema>;

/**
 * S2.2 / Phase 1 trust correction: ``signature_verification_status``
 * distinguishes ``verified`` (HMAC matched), ``failed`` (mismatch or
 * non-active status), and ``unavailable`` (verify could not run, e.g. no
 * secret).  Newer UIs MUST consume the status string when rendering a
 * verification badge; the legacy ``signature_verified`` boolean is kept
 * for back-compat (true only for ``verified``).
 */
export const SignatureVerificationStatusSchema = z.enum([
  "verified",
  "failed",
  "unavailable",
]);
export type SignatureVerificationStatus = z.infer<
  typeof SignatureVerificationStatusSchema
>;

export const AgentCardSchema = z.object({
  agent_id: z.string(),
  agent_name: z.string(),
  owner: z.string(),
  version: z.string(),
  description: z.string(),
  capabilities: z.array(CapabilitySchema),
  policies: z.array(PolicySchema),
  status: z.string(),
  valid_until: z.string().datetime({ offset: true }).nullable(),
  parent_agent_id: z.string().nullable(),
  signature_truncated: z.string(),
  signature_verified: z.boolean(),
  signature_verification_status: SignatureVerificationStatusSchema,
  created_at: z.string().datetime({ offset: true }),
  updated_at: z.string().datetime({ offset: true }),
});
export type AgentCard = z.infer<typeof AgentCardSchema>;

export const AgentCardListSchema = z.array(AgentCardSchema);
export type AgentCardList = z.infer<typeof AgentCardListSchema>;

export const AgentAuditEntrySchema = z.object({
  agent_id: z.string(),
  action: z.string(),
  performed_by: z.string(),
  timestamp: z.string().datetime({ offset: true }),
  details: z.record(z.unknown()),
});
export type AgentAuditEntry = z.infer<typeof AgentAuditEntrySchema>;

export const AgentAuditEntryListSchema = z.array(AgentAuditEntrySchema);
export type AgentAuditEntryList = z.infer<typeof AgentAuditEntryListSchema>;

export const IntegrityReportSchema = z.object({
  workflow_id: z.string(),
  chain_valid: z.boolean(),
  broken_at_event_id: z.string().nullable(),
  expected_hash: z.string().nullable(),
  actual_hash: z.string().nullable(),
});
export type IntegrityReport = z.infer<typeof IntegrityReportSchema>;

export const CorrelationHealthSchema = z.object({
  has_trace_id: z.boolean(),
  has_user_id: z.boolean(),
  has_task_id: z.boolean(),
  has_agent_id: z.boolean(),
  missing_keys: z.array(z.string()),
});
export type CorrelationHealth = z.infer<typeof CorrelationHealthSchema>;

export const ComplianceBundleSchema = z.object({
  workflow_id: z.string(),
  event_count: z.number().int().nonnegative(),
  hash_chain_valid: z.boolean(),
  bundle_type: z.string(),
  exported_at: z.string().datetime({ offset: true }).nullable(),
  events: z.array(BlackBoxEventSchema),
  identity_cards: z.record(AgentCardSchema.nullable()),
  audit_trails: z.record(z.array(AgentAuditEntrySchema)),
  phase_decisions: z.array(DecisionRecordSchema),
  correlation_health: CorrelationHealthSchema,
  /**
   * Sprint 3 review F4: bundle now embeds the full IntegrityReport so the
   * Workflow Deep Dive Recording quadrant can name the broken event id.
   */
  integrity: IntegrityReportSchema,
});
export type ComplianceBundle = z.infer<typeof ComplianceBundleSchema>;

/**
 * Sprint 3 review F3 fix: batched compliance summary contract.
 *
 * Replaces the per-row N+1 fan-out where the page used to call
 * `getWorkflowIntegrity` once per workflow.  One network round-trip now
 * returns every workflow + integrity row, plus the bounds the server
 * actually applied.
 */
export const WorkflowIntegritySummarySchema = z.object({
  workflow: WorkflowSummarySchema,
  integrity: IntegrityReportSchema.nullable(),
});
export type WorkflowIntegritySummary = z.infer<
  typeof WorkflowIntegritySummarySchema
>;

export const ComplianceSummarySchema = z.object({
  rows: z.array(WorkflowIntegritySummarySchema),
  generated_at: z.string().datetime({ offset: true }),
  since: z.string().datetime({ offset: true }).nullable(),
  until: z.string().datetime({ offset: true }).nullable(),
});
export type ComplianceSummary = z.infer<typeof ComplianceSummarySchema>;

export const LogRowSchema = z.object({
  concern: z.string(),
  timestamp: z.string().datetime({ offset: true }).nullable(),
  logger: z.string(),
  level: z.string(),
  message: z.string(),
  raw: z.string(),
});
export type LogRow = z.infer<typeof LogRowSchema>;

export const LogRowListSchema = z.array(LogRowSchema);
export type LogRowList = z.infer<typeof LogRowListSchema>;

/**
 * Stable list of log concern keys -- mirror of `DEFAULT_LOG_CONCERNS` in
 * `services/explainability_service.py`.  Adding a new handler in
 * `logging.json` requires a new entry here AND a matching entry on the
 * Python side.
 */
export const LOG_CONCERN_KEYS = [
  "prompts",
  "guards",
  "evals",
  "tools",
  "routing",
  "black_box",
  "phases",
  "identity",
  "drift",
  "framework_telemetry",
  "trust_trace",
  "authorization",
  "long_term_memory",
  "agent_ui_adapter_server",
  "agent_ui_adapter_transport",
  "agent_ui_adapter_translators",
  "explainability",
] as const;

export type LogConcern = (typeof LOG_CONCERN_KEYS)[number];

export const LOG_LEVELS = ["INFO", "WARN", "ERROR"] as const;
export type LogLevel = (typeof LOG_LEVELS)[number];

export const HealthResponseSchema = z.object({
  status: z.string(),
});
export type HealthResponse = z.infer<typeof HealthResponseSchema>;

export const ErrorResponseSchema = z.object({
  detail: z.string(),
});
export type ErrorResponse = z.infer<typeof ErrorResponseSchema>;
